Remove repetitive integration tests

lalexgap opened the pull request:
Implements a new integration test that accepts a configurable test case. The goal of this integration test is to replace the existing collection of repetitive test with a more comprehensive test.

I've introduced a new `TestCase` struct that specifies the various parameters for a run of an integration test:
```
// TestCase is a test case for the client integration test.
type TestCase struct {
	Description    string
	Chain          ChainType
	MessageService MessageService
	NumOfChannels  uint
	NumOfPayments  uint
	MessageDelay   time.Duration
	LogName        string
	NumOfHops      uint
	Participants   []TestParticipant
}

```


**NOTE:** This is based on https://github.com/statechannels/go-nitro/pull/1201  and will fail until that PR (and the accompany testground PR) is merged in.

I've created some issues for remaining work or issues to solve around this new integration test:
- https://github.com/statechannels/go-nitro/issues/1204
- https://github.com/statechannels/go-nitro/issues/1203
- https://github.com/statechannels/go-nitro/issues/1202

lalexgap commented on client/engine/messageservice/p2p-message-service/service.go:
```diff
@@ -179,3 +173,35 @@ func (s *P2PMessageService) Close() error {
 }
+
+// generateKey generates a ECDSA key deterministically from the given address
+// TOODO: This is probably a security hole, but it's good enough for now
```
This makes it pretty trivial to generate the message key if you know the address. For our use case this is fine (since we don't even care if the message traffic is encrypted, since we verify signatures about things we care about).

Still, probably better to move this to test code and just have the message service accept an `ecdsa.PrivKey` for it's message service. In the real world we can generate one randomly but in our tests we can use a deterministic set of keys.

kerzhner replied:
It seems like, before this PR, we are generating the p2p message key from the state channels private key. Is there a reason we are moving away from this approach?

lalexgap replied:
> It seems like, before this PR, we are generating the p2p message key from the state channels private key. Is there a reason we are moving away from this approach?

FYI, I ended up tackling some of this in a separate [PR](https://github.com/statechannels/go-nitro/pull/1194), so this function is outdated. In that PR `generateMessageKey` accepts a `pk` not the address.

lalexgap commented on client_test/integration/cases.go:
```diff
@@ -0,0 +1,21 @@
+
+import "github.com/statechannels/go-nitro/internal/testactors"
+
+var simpleCase = TestCase{
```
TODO: Add some simple validation that makes sure all test cases values are "valid". For example, right now we only support up to 2 hops.

lalexgap commented on client/engine/messageservice/p2p-message-service/service.go:
```diff
@@ -51,13 +49,17 @@ func (ms *P2PMessageService) Id() peer.ID {
 	return id
 }
 
+func (ms *P2PMessageService) PeerInfo() PeerInfo {
```
I don't think we need this anymore.

lalexgap commented on client_test/integration/types.go:
```diff
@@ -0,0 +1,58 @@
+	Participants   []TestParticipant
+}
+
+type sharedInra struct {
```
Rename this

kerzhner commented on client/engine/messageservice/p2p-message-service/service.go:
```diff
@@ -179,3 +173,35 @@ func (s *P2PMessageService) Close() error {
+// generateKey generates a ECDSA key deterministically from the given address
+// TOODO: This is probably a security hole, but it's good enough for now
+func generateKey(address common.Address) p2pcrypto.PrivKey {
+
```
Nit
```suggestion
```

kerzhner commented on client/engine/messageservice/p2p-message-service/service.go:
```diff
@@ -179,3 +173,35 @@ func (s *P2PMessageService) Close() error {
+
+	messageKey, _, err := p2pcrypto.GenerateECDSAKeyPair(createReader(address))
+	if err != nil {
+		panic(err)
```
Should we bubble up these errors?

lalexgap replied:
Yup! In general I need to do a pass on this PR to clean up error handling. I've just been using `panic`s so I don't miss errors while building/debugging.

kerzhner commented on client/engine/messageservice/p2p-message-service/service.go:
```diff
@@ -179,3 +173,35 @@ func (s *P2PMessageService) Close() error {
+		panic(err)
+	}
+
+	if err != nil {
+		panic(err)
+	}
```
```suggestion
```

lalexgap commented on client_test/integration/integration_test.go:
```diff
@@ -0,0 +1,105 @@
+			aliceLedgers := make([]types.Destination, tc.NumOfHops)
+			bobLedgers := make([]types.Destination, tc.NumOfHops)
+			for i, clientI := range intermediaries {
+				aliceLedgers[i] = setupLedgerChannel(t, clientA, clientI, common.Address{})
```
We need to add some correctness checks in here.

lalexgap commented on client_test/helpers_test.go:
```diff
@@ -182,7 +337,7 @@ func checkPaymentChannel(t *testing.T, id types.Destination, o outcome.Exit, sta
 		}
 		if diff := cmp.Diff(expected, ledger, cmp.AllowUnexported(big.Int{})); diff != "" {
-			t.Fatalf("Payment channel diff mismatch (-want +got):\n%s", diff)
+			panic(fmt.Errorf("payment channel diff mismatch (-want +got):\n%s", diff))
```
Happy to revert this change, but with the `panic` I was able to get a full stack trace and figure out which call to `checkPaymentChannel` was failing. With `t.Fatalf` I don't think we get any information about who called `checkPaymentChannel`

kerzhner commented on client_test/integration_test.go:
```diff
@@ -0,0 +1,178 @@
+				intermediaryAddresses = append(intermediaryAddresses, *intermediaries[1].Address)
+			}
+
+			defer clientA.Close()
```
It might be clearer to defer client close calls immediately below each client instantiation. In other words:
```
clientA := ...
defer clientA.Close()
```

lalexgap replied:
Updated in https://github.com/statechannels/go-nitro/pull/1193/commits/f2f6009fe5eacd03af2c2c97797b45614a78da75

kerzhner commented on client_test/types_test.go:
```diff
@@ -0,0 +1,91 @@
+
+type TestParticipant struct {
+	StoreType StoreType
+	Name      testactors.ActorName
```
An alternative is for `TestParticipant` to contain the `testactors.Actor` (instead of just the name). That way we don't have "a foreign key" relationship from a TestParticipant to an actor.

lalexgap replied:
Good idea 👍

lalexgap replied:
Updated in https://github.com/statechannels/go-nitro/pull/1193/commits/30bf516375af1918c2a983fc2b67d726df9ac6b9

kerzhner commented on client_test/types_test.go:
```diff
@@ -0,0 +1,91 @@
+		tc.NumOfHops == 2 && len(tc.Participants) != 4 {
+		return fmt.Errorf("NumOfHops is %d, but there are %d participants", tc.NumOfHops, len(tc.Participants))
+	}
+	if tc.NumOfChannels < 1 {
+		return fmt.Errorf("NumOfChannels must be greater than 0")
+	}
+	if tc.NumOfChannels > 10 {
+		return fmt.Errorf("NumOfChannels must be smaller than 10")
+	}
```
```suggestion
	if tc.NumOfChannels < 1 ||  tc.NumOfChannels > 9{
		return fmt.Errorf("NumOfChannels must be greater than 0 and less than 10. Supplied %d", tc.NumOfChannels)
	}
```

kerzhner commented on client_test/helpers_test.go:
```diff
@@ -170,6 +116,217 @@ func newLogWriter(logFile string) *os.File {
+		broker := messageservice.NewBroker()
+		infra.broker = &broker
+	case P2PMessageService:
+
```
```suggestion
```

kerzhner commented on client_test/integration_test.go:
```diff
@@ -0,0 +1,178 @@
+			clientB := setupIntegrationClient(tc, testactors.BobName, infra)
+
+			intermediaries := []client.Client{setupIntegrationClient(tc, testactors.IreneName, infra)}
+			intermediaryAddresses := []types.Address{*intermediaries[0].Address}
```
Seems like we can construct this array from `intermediaries` when we need the address list. This logic is a bit fragile since it's easy to forget to add to the `intermediaryAddresses` list when a new intermediary is added to `intermediaries`.

lalexgap replied:
Yup good point!

kerzhner commented on client_test/integration_test.go:
```diff
@@ -0,0 +1,178 @@
+			closeLedgerChannel(t, clientA, intermediaries[0], aliceLedgers[0])
+			checkLedgerChannel(t, aliceLedgers[0], finalAliceLedger(*intermediaries[0].Address, asset, tc.NumOfPayments, 1, tc.NumOfChannels), query.Complete, &clientA)
+
+			// TODO: This is brittle, we should generalize this to
```
Seems like an incomplete comment

kerzhner commented on client_test/integration_test.go:
```diff
@@ -0,0 +1,178 @@
+				<-clientB.ReceivedVouchers()
+			}
+
+			// TODO: This check sometime flickers
```
Is there an example of this flicker? How does the test fail?

lalexgap replied:
This turned out to be a self-caused flicker 🤦 We were waiting for `numVirtualChannel` vouchers. However if we're making more than 1 payment per channel we wouldn't be waiting for all the vouchers. Updated in https://github.com/statechannels/go-nitro/pull/1193/commits/e9ea072f974694ad94d5945f3847e4851cc88f77

kerzhner commented on client_test/helpers_test.go:
```diff
@@ -170,6 +116,217 @@ func newLogWriter(logFile string) *os.File {
+	return testdata.Outcomes.Create(alpha, beta, virtualChannelDeposit, 0, asset)
+}
+
+// TODO: This is currently unused as a the final payment check seems to flic
```
```suggestion
// TODO: This is currently unused as a the final payment check seems to flicker
```

kerzhner commented on client_test/helpers_test.go:
```diff
@@ -170,6 +116,217 @@ func newLogWriter(logFile string) *os.File {
+
+		<-b.ObjectiveCompleteChan(objectiveId)
+
+		// TODO: For some reason we're not beeing an ojective complete event from the second intermediary
```
```suggestion
		// TODO: For some reason we're not seeing an objective complete event from the second intermediary
```

kerzhner reviewed (commented):
Thanks for putting together this PR! It's a big improvement over the existing smattering of client tests that duplicate setup and the code tested.

It seems like there are unresolved issues with the integration test including flickering and not receiving an objective. Are these issues similar to what we see in the existing testsuite?

The existing unit testsuite runs fairly reliably on CI. There are local failures when tests are run repeatedly (over 10 times). Do we think we should increase the reliability of the new integration test before merging the PR? Or do we think the integration test is reliable enough?

kerzhner commented on client_test/integration_test.go:
```diff
@@ -0,0 +1,178 @@
+	// Clean up all the test data we create at the end of the test
+	defer os.RemoveAll(STORE_TEST_DATA_FOLDER)
+
+	for _, tc := range cases {
```
While we have a small number of test cases, it might be nice to spit every test case into a separate function so that we can run the test case individually. It would make the debugging experience nicer.

lalexgap replied:
Updated in https://github.com/statechannels/go-nitro/pull/1193/commits/ab2ca01266ba8a86d0687dcadc6a72418111dc0a

lalexgap commented on client_test/helpers_test.go:
```diff
@@ -170,6 +116,217 @@ func newLogWriter(logFile string) *os.File {
+		logDestination := newLogWriter(tc.LogName)
+
+		ethAcountIndex := a.Port - testactors.START_PORT
+		cs, err := chainservice.NewSimulatedBackendChainService(*si.simulatedChain, *si.bindings, si.ethAccounts[ethAcountIndex], logDestination)
```
Good catch! Updated in https://github.com/statechannels/go-nitro/pull/1193/commits/afc5641c27c442c5e3a1de6000057330a9588512

lalexgap commented:
> It seems like there are unresolved issues with the integration test including flickering and not receiving an objective. Are these issues similar to what we see in the existing testsuite?

I think I've now removed the TODOs around flickering and not receiving an objective:
- The payment channel query was flickering because I wasn't waiting for the correct amount vouchers. This meant that we queried the channel too early, explaining why the totals weren't correct. Fixed in https://github.com/statechannels/go-nitro/pull/1193/commits/e9ea072f974694ad94d5945f3847e4851cc88f77
- Both intermediaries seem to be receiving an objective, so I've re-enabled checking all intermediaries for an objective in https://github.com/statechannels/go-nitro/pull/1193/commits/9ebd0ec0951b0da3497178029741c2180279943f.  I probably misdiagnosed the original issue when I disabled the check.

> The existing unit testsuite runs fairly reliably on CI. There are local failures when tests are run repeatedly (over 10 times). Do we think we should increase the reliability of the new integration test before merging the PR? Or do we think the integration test is reliable enough?

With the recent changes I now think the integration test is pretty stable (or at least on my machine 😉 ). Locally I'm able to run all the tests 100 times:
```
go test ./... --shuffle on --count 100 --timeout 10m
?   	github.com/statechannels/go-nitro/abi	[no test files]
?   	github.com/statechannels/go-nitro/client	[no test files]
?   	github.com/statechannels/go-nitro/client/engine	[no test files]
ok  	github.com/statechannels/go-nitro/channel	0.724s
?   	github.com/statechannels/go-nitro/client/engine/chainservice/consensusapp	[no test files]
?   	github.com/statechannels/go-nitro/client/engine/chainservice/erc20	[no test files]
?   	github.com/statechannels/go-nitro/client/engine/chainservice/virtualpaymentapp	[no test files]
?   	github.com/statechannels/go-nitro/client/engine/messageservice/p2p-message-service	[no test files]
?   	github.com/statechannels/go-nitro/client/query	[no test files]
?   	github.com/statechannels/go-nitro/crypto	[no test files]
?   	github.com/statechannels/go-nitro/internal/logging	[no test files]
?   	github.com/statechannels/go-nitro/internal/safesync	[no test files]
?   	github.com/statechannels/go-nitro/internal/testactors	[no test files]
?   	github.com/statechannels/go-nitro/internal/testhelpers	[no test files]
ok  	github.com/statechannels/go-nitro/channel/consensus_channel	5.590s
ok  	github.com/statechannels/go-nitro/channel/state	0.521s
ok  	github.com/statechannels/go-nitro/channel/state/outcome	0.567s
ok  	github.com/statechannels/go-nitro/client/engine/chainservice	2.779s
?   	github.com/statechannels/go-nitro/rand	[no test files]
?   	github.com/statechannels/go-nitro/rpc/transport	[no test files]
?   	github.com/statechannels/go-nitro/rpc/transport/nats	[no test files]
?   	github.com/statechannels/go-nitro/rpc/transport/ws	[no test files]
ok  	github.com/statechannels/go-nitro/client/engine/chainservice/adjudicator	36.411s
ok  	github.com/statechannels/go-nitro/client/engine/chainservice/create2deployer	1.921s
ok  	github.com/statechannels/go-nitro/client/engine/messageservice	0.469s
ok  	github.com/statechannels/go-nitro/client/engine/store	0.685s
ok  	github.com/statechannels/go-nitro/client_test	45.927s
ok  	github.com/statechannels/go-nitro/payments	0.332s
ok  	github.com/statechannels/go-nitro/protocols	0.188s
ok  	github.com/statechannels/go-nitro/protocols/directdefund	2.244s
ok  	github.com/statechannels/go-nitro/protocols/directfund	1.304s
ok  	github.com/statechannels/go-nitro/protocols/virtualdefund	2.789s
ok  	github.com/statechannels/go-nitro/protocols/virtualfund	10.483s
ok  	github.com/statechannels/go-nitro/rpc	0.252s
ok  	github.com/statechannels/go-nitro/rpc/serde	0.269s
ok  	github.com/statechannels/go-nitro/types	0.153s
```

kerzhner reviewed (approved):
Looks great, thanks for putting this PR together!
